tokio-pty-process/src/lib.rs

```diff
@@ -135,6 +141,14 @@ impl AsyncPtyMaster {
                 return Err(io::Error::last_os_error());
             }
 
+            #[cfg(target_os = "freebsd")]
+            {
+                let flags = libc::fcntl(fd, libc::F_GETFL, 0);
+                if libc::fcntl(fd, libc::F_SETFL, flags | libc::O_NONBLOCK) == -1 {
+                    return Err(io::Error::last_os_error());
+                };
+            }
```

Comment:
Funnily enough, if this hunk is removed, the following still does not panic on FreeBSD:
```rust
extern crate tokio;
extern crate tokio_pty_process;

use tokio::prelude::*;
use tokio_pty_process::AsyncPtyMaster;

struct AsyncReadTest(AsyncPtyMaster);

impl Future for AsyncReadTest {
    type Item = ();
    type Error = std::io::Error;

    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
        let mut buf = vec![0u8, 100];
        self.0.read(&mut buf).map(|_| Async::Ready(()))
    }
}

fn main() {
    let err = AsyncPtyMaster::open()
        .map(AsyncReadTest)
        .expect("could not open pty")
        .wait()
        .expect_err("Read succeeded but no data available");

    assert_eq!(err.kind(), std::io::ErrorKind::WouldBlock);
}
```

Can you confirm this behavior on Linux too, if the AsyncPtyMaster is not opened with `O_NONBLOCK`?

Reply:
Yes, I get this behavior ... but if I strace the program, it isn't actually doing any PTY I/O! I am not sure what's going on — for some reason I can't step through it in GDB at the moment.

Reply:
You're right, I see that too:
```
$ kdump -i ktrace.out
[...]
 34900 tokio-runtime-worke RET   thr_new 0
 34900 wouldblock RET   fork 0
 34900 wouldblock CALL  sigaltstack(0,0x7fffdfdfcf80)
 34900 wouldblock RET   sigaltstack 0
 34900 wouldblock CALL  mmap(0,0x8800,0x3<PROT_READ|PROT_WRITE>,0x1002<MAP_PRIVATE|MAP_ANON>,0xffffffff,0)
 34900 tokio-runtime-worke CALL  posix_openpt(0x8002)
 34900 wouldblock RET   mmap 34384453632/0x801792000
 34900 wouldblock CALL  sigaltstack(0x7fffdfdfcf80,0)
 34900 wouldblock RET   sigaltstack 0
 34900 tokio-runtime-worke RET   posix_openpt 15/0xf
 34900 wouldblock CALL  thr_set_name(0x18a46,0x80177d0e0)
[...]
```